syweb/webclient/app/components/matrix/notification-service.js

```diff
@@ -48,23 +49,293 @@ angular.module('notificationService', ['matrixService'])
+                // notify when you are invited
+                message = displayname + " invited you to a room";
+            }
+        }
+        return message;
+    };
+
+    var matchingRuleForEventWithRulesets = function(ev, rulesets) {
+        if (!rulesets) return null;
+        if (ev.user_id == matrixService.config().user_id) return null;
+        angular.forEach(rulesets.device, function(devname, devrules) {
+            var matchingRule = matchingRuleFromKindSet(devrules, devname);
+            if (matchingRule) defer.resolve(matchingRule);
+        });
+        return matchingRuleFromKindSet(ev, rulesets.global);
+    };
+
+    var matchingRuleForEventNow = function(ev) {
+        // Returns a matching rule for this event with whatever rules
+        // we have cached right now (ie. it might be total rubbish)
+        return matchingRuleForEventWithRulesets(ev, rulesCache);
+    };
+
+    var shouldHighlightEventWithRule = function(ev, rule) {
+        var actionObj = actionListToActionsObject(rule.actions);
+        if (!actionObj.notify) return false;
+
+        if (actionObj.tweaks.highlight == undefined) {
+            // if it isn't specified, we highlight
+            return true;
+        }
+
+        return !!actionObj.tweaks.highlight;
+    };
+
+    var actionListToActionsObject = function(actionlist) {
+        var actionobj = { 'notify': false, 'tweaks': {} };
+        angular.forEach(actionlist, function(action) {
+            if (action === 'notify') {
+                actionobj.notify = true;
+            } else if (typeof action === 'object') {
+                if (action.value == undefined) action.value = true;
+                actionobj.tweaks[action.set_tweak] = action.value;
+            }
+        });
+        return actionobj;
+    };
+
+    var showNotification = function(title, body, icon, onclick, tag, audioNotification) {
+        if (!tag) {
+            tag = "matrix";
+        }
+        var notification = new window.Notification(
+            title,
+            {
+                "body": body,
+                "icon": icon,
+                "tag": tag
+            }
+        );
+
+        if (onclick) {
+            notification.onclick = onclick;
+        }
+        
+        var audioClip;
+        
+        if (audioNotification) {
+            audioClip = playAudio(audioNotification);
+        }
+
+        $timeout(function() {
+            notification.close();
+            if (audioClip) {
+                audioClip.pause();
+            }
+        }, 5 * 1000);
+    };
     
-    return {
+
+    var nfsvc = {
 
         clearRulesCache : function() {
-            rulesCache = null;
+            rulesCacheIsCurrent = false;
         },
 
-        getGlobalRulesets : function() {
+        getRulesets : function() {
             var def = $q.defer();
 
-            if (!rulesCache && !ruleFetchPromise)  {
+            if (!rulesCacheIsCurrent && !ruleFetchPromise)  {
```

Comment:
If I understand ``ruleFetchPromise`` correctly, you're using it to prevent ``getRulesets`` making multiple HTTP calls whilst there is still a request outstanding. You're using ``rulesCacheIsCurrent`` to allow public API calls to force a refresh of this HTTP call. 

This can be done in a slightly cleaner way by using the fact that promises store their success state. An implementation of this which is almost identical to this behaviour can be seen in ``event-handler-service.js`` in ``initialSyncDeferred`` and ``reset()``.

Simply put, if a deferred is already ``resolve()``d, you can still attach ``then(..)`` to the promise and it will be invoked with success/failure on the next digest cycle with the success/failure state. This means that you could do something to the effect of:
```
var getRulesetsPromise = null;

clearRulesCache: function() {
  getRulesetsPromise = null;
}
getRulesets: function() {
  if (getRulesetsPromise == null) {
    // do the HTTP call and set the promise
    getRulesetsPromise = matrixService.getPushRules()
  }
  return getRulesetsPromise;
}
```
Doing this would make this logic easier to grok, as well as remove the need for additional state variables.